src/librustc_resolve/lib.rs

```diff
@@ -172,13 +172,90 @@ fn resolve_struct_error<'sess, 'a>(resolver: &'sess Resolver,
                                    resolution_error: ResolutionError<'a>)
                                    -> DiagnosticBuilder<'sess> {
     match resolution_error {
-        ResolutionError::TypeParametersFromOuterFunction => {
+        ResolutionError::TypeParametersFromOuterFunction(outer_def) => {
             let mut err = struct_span_err!(resolver.session,
                                            span,
                                            E0401,
-                                           "can't use type parameters from outer function; \
-                                           try using a local type parameter instead");
+                                           "can't use type parameters from outer function");
             err.span_label(span, "use of type variable from outer function");
+            match outer_def {
+                Def::SelfTy(_, maybe_impl_defid) => {
+                    if let Some(impl_span) = maybe_impl_defid.map_or(None,
+                            |def_id| resolver.definitions.opt_span(def_id)) {
+                        // Adjust the span so that just the `impl` keyword is taken by removing
+                        // everything after `<` (`"impl<T> Iterator for A<T> {}" -> "impl"`) and
+                        // everything after the first whitespace (`"impl Iterator for A" -> "impl"`)
+                        let cm = resolver.session.codemap();
+                        let impl_span = cm.span_until_char(impl_span, '<');
+                        let impl_span = cm.span_until_whitespace(impl_span);
```

Comment:
whoosh, this kind of manipulation feels very fragile (I strongly suspect we are overlooking cases here). Probably ok for this PR, but we should think about saving these spans during the parsing stage or something.

Reply:
At minimum, can we move these adjustments into some sort of helpers, so that we can tweak them over time and share them between error reporting code?

Reply:
I can put some helper functions in `codemap.rs` for these manipulations, but as you say they are fragile and putting them here may encourage to use them. I think that if we make them easily acccessible, then we should transition them to more robust versions as soon as possible. I don't know how to save the spans that we want, but maybe the functions in codemap could work on parsed tokens instead of string? (I'm thinking of the `quote` crate used for the procedural macros, but maybe there's another way of doing that since we're inside the `syntax` crate)

Reply:
I'm not sure ... let's land it like this for now I guess ...